core/test_network/src/unique_port.rs

```diff
@@ -0,0 +1,65 @@
+#![allow(clippy::restriction)]
+
+use once_cell::sync::Lazy;
+use std::net::{Ipv4Addr, SocketAddrV4, TcpListener};
+use std::ops::Range;
+use std::sync::Mutex;
+
+static PORT_IDX: Lazy<Mutex<u16>> = Lazy::new(|| Mutex::new(1000));
+
+#[macro_export]
+macro_rules! generate_offset {
+    () => {{
+        use std::collections::hash_map::DefaultHasher;
+        use std::hash::{Hash, Hasher};
+
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str {
+            std::any::type_name::<T>()
+        }
+        let name = type_name_of(f);
+        let name = &name[..name.len() - 3];
+        let mut hasher = DefaultHasher::new();
+        name.hash(&mut hasher);
+        1000 + (hasher.finish() % ((u16::MAX - 1000) as u64)) as u16
```

Comment:
I assume that the casts here are intentional. It's not obvious even to me, why the numbers are the way they are. 

Why isn't it, for example:
```suggestion
        (hasher.finish() % (u16::MAX as u64) as u16
```
I would add comments.

Reply:
Yeah, I'll document everything before I open the PR.